### ISSUE TITLE
[RFC] Proposed Changes to Feature Tracking & Classification for PyTorch Releases starting Release 2.8

### ISSUE THREAD

**Issue by atalman** [pytorch/pytorch#152134]
### 🚀 The feature, motivation and pitch

RFC Authors: @anitakat @atalman 

Hello everyone, 

Following feedback and discussion on existing gaps of the feature review process, below are proposed changes for which we are keen to have your input.  

## Feature Tracking Process
Beginning with release 2.8, the PyTorch release will only track major features. At this time we do not have a comprehensive list and welcome examples from the community of what they would like tracked. 
* All major features will require an RFC from the start with an estimated timeline. This will allow the maintainers to provide async feedback before feature implementation begins.  
* The RFC for these major features will have a current status that will enable partners and the community at large, to reference the progress of a given feature with ease. [Example of an RFC](https://github.com/pytorch/pytorch/issues/130249).
* These RFCs will be tagged and labelled appropriately for tracking and when a feature is complete/stable, they will be untagged and no longer tracked. 
* Release notes will highlight new major features and improvements made to existing features, and will follow the new classification of API-Stable and API-Unstable. 
For any features not on this list, the only requirement is to follow the path to stable below, to be classified as stable when ready. 

## Feature Classification:
Beginning with release 2.8, feature submissions will be classified as either API-Stable or API-Unstable, and the previous classifications of Prototype, Beta and Stable, will no longer be used. 

### API-Stable
(Previously called Stable) An API-Stable feature means that the user value-add has been proven, the API isn’t expected to change, the feature is performant and all documentation exists to support end user adoption.

Examples of API-Stable features include Accelerated Transformers, DataLoader, Autograd Engine, and CUDA support in PyTorch CI/CD. 

Commitment from the PyTorch team: We expect to maintain these features long term and generally there should be no major performance limitations or gaps in documentation. We also expect to maintain backwards compatibility (although breaking changes can happen and notice will be given one release ahead of time).
### API-Unstable
(Previously Prototype or Beta)  Encompasses all features that are under active development where API may change based on user feedback, requisite performance improvements or because coverage across operators is not yet complete. 

Commitment from the PyTorch team: We are not committing to backwards compatibility. The APIs and performance characteristics of this feature may change.

### Classification Requirements

Requirement | API-Unstable | API-Stable | Path to API-Stable
-- | -- | -- | --
RFC Created | X | X | -
Doc Strings | X | X | -
Unit Tests | X | X | -
CI Coverage | X | X | -
Complete Workflow Coverage (e.g. CV or NLP) |   | X | Phase 1
Recipe or Tutorial |   | X | Phase 1
User Feedback (Features with User API surface) |   | X | Phase 2
Dogfooding: 1-2 early adopter teams (internal or external) have found this feature useful and their feedback has been incorporated |   | X | Phase 2
Design review / TL Signoff |   | X | Phase 2
API Stability |   | X | -
Full Op Coverage |   | X | -

### Path To Stable API

<img width="778" alt="Image" src="https://github.com/user-attachments/assets/e9ae8fcc-6d4c-41fe-bcb6-26330f82fdd8" />

Thank you for reading and we look forward to your feedback. 

Cheers, 

Team PyTorch

**Comment by seemethere** [comment]
I think for things that do not have a public facing API we can denote them with:

* Stable
* Unstable

This is for things like Blackwell upgrades, Python support, etc.

**Comment by Xia-Weiwen** [comment]
Hi @atalman How should I determine performance optimization to be stable or unstable? The optimization is not user-facing. Thanks.

**Comment by atalman** [comment]
Hi @Xia-Weiwen the perf optimization is significant enough and its something you want to include in Release Blog please go ahead and open a "Release highlight for proposed Feature" issue.

**Comment by dvrogozh** [comment]
It's proposed to create an RFC issue for each feature which will be tracked. Further, "Release highlight for proposed Feature" issue template is mentioned. Can you, please, clarify how "Release highlight <...>" issue is related to original RFC issue proposed to be created? is it actually meant to be the original RFC issue? or "Release highlight <...>" is a special issue to request description added to particular release notes/blog post?

Also, if the feature is a complex one and can be broken into independent parts, should there be multiple RFCs or single one? should there be a single wrapping RFC with multiple subitems? One such example might be breaking feature implementation for Linux and Windows.

**Comment by atalman** [comment]
HI @dvrogozh I believe the "Release highlight for proposed Feature" - should be both  : original RFC issue proposed to be created and also serve as a description for particular release notes/blog post.

For complex features, I believe we probably would be Looking for 1 RFC/Release highlight as umbrella issue to which we can attach multiple subitems, which can also be RFCs themselves however probably not necessary labeled with ``release-feature-request`` label

**Comment by malfet** [comment]
> Complete Workflow Coverage (e.g. CV or NLP)

Just curious, what does that mean/entail?

**Comment by atalman** [comment]
Hi @malfet Good Call this should be ``Complete CI Coverage`` this was taken from one of the earlier documents specifying PyTorch Release process

**Comment by atalman** [comment]
Hi @malfet I believe this must be copy/paste issue from one of the earliest docs. 
Template that is used in the request for feature here for example: https://dev-discuss.pytorch.org/t/pytorch-release-2-6-0-call-for-features/2574 do not have this condition. I think ``Complete Workflow Coverage = CI Coverage`` hence removing this condition.

**Comment by johnnynunez** [comment]
Plan CUDA 13 support? even if it is in the nightly branch.
 I mean, there is a restructured codegens for blackwell and new codegen devices.
Also better support for SBSA. I hope that the release come this month

**Comment by fanta1mohamed** [comment]
> RFC Authors: [@anitakat](https://github.com/anitakat) [@atalman](https://github.com/atalman)
> 
> Hello everyone,
> 
> Following feedback and discussion on existing gaps of the feature review process, below are proposed changes for which we are keen to have your input.
> 
> ## Feature Tracking Process
> Beginning with release 2.8, the PyTorch release will only track major features. At this time we do not have a comprehensive list and welcome examples from the community of what they would like tracked.
> 
> * All major features will require an RFC from the start with an estimated timeline. This will allow the maintainers to provide async feedback before feature implementation begins.
> * The RFC for these major features will have a current status that will enable partners and the community at large, to reference the progress of a given feature with ease. [Example of an RFC](https://github.com/pytorch/pytorch/issues/130249).
> * These RFCs will be tagged and labelled appropriately for tracking and when a feature is complete/stable, they will be untagged and no longer tracked.
> * Release notes will highlight new major features and improvements made to existing features, and will follow the new classification of Stable (API-Stable) and Unstable (API-Unstable).
>   For any features not on this list, the only requirement is to follow the path to stable below, to be classified as stable when ready.
> 
> ## Feature Classification:
> Beginning with release 2.8, feature submissions will be classified as either Stable (API-Stable) or Unstable (API-Unstable), and the previous classifications of Prototype, Beta and Stable, will no longer be used.
> 
> ### Stable (API-Stable)
> (Previously called Stable) An API-Stable feature means that the user value-add has been proven, the API isn’t expected to change, the feature is performant and all documentation exists to support end user adoption.
> 
> Examples of API-Stable features include Accelerated Transformers, DataLoader, Autograd Engine, and CUDA support in PyTorch CI/CD.
> 
> Level of Commitment: We expect to maintain these features long term and generally there should be no major performance limitations or gaps in documentation. We also expect to maintain backwards compatibility (although breaking changes can happen and notice will be given one release ahead of time).
> 
> ### Unstable (API-Unstable)
> (Previously Prototype or Beta) Encompasses all features that are under active development where API may change based on user feedback, requisite performance improvements or because coverage across operators is not yet complete.
> 
> Level of Commitment: We are not committing to backwards compatibility. The APIs and performance characteristics of this feature may change.
> 
> ### Classification Requirements
> Requirement	Unstable (API-Unstable)	Stable (API-Stable)	Path to Stable (API-Stable)
> RFC Created	X	X	-
> Doc Strings	X	X	-
> Unit Tests	X	X	-
> CI Coverage	X	X	-
> Recipe or Tutorial	 	X	Phase 1
> User Feedback (Features with User API surface)	 	X	Phase 2
> Dogfooding: 1-2 early adopter teams (internal or external) have found this feature useful and their feedback has been incorporated	 	X	Phase 2
> Design review / TL Signoff	 	X	Phase 2
> API Stability	 	X	-
> Full Op Coverage	 	X	-
> ### Path To Stable API
> <img alt="Image" width="778" src="https://private-user-images.githubusercontent.com/7563158/437171116-e9ae8fcc-6d4c-41fe-bcb6-26330f82fdd8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM3NjM2OTYsIm5iZiI6MTc1Mzc2MzM5NiwicGF0aCI6Ii83NTYzMTU4LzQzNzE3MTExNi1lOWFlOGZjYy02ZDRjLTQxZmUtYmNiNi0yNjMzMGY4MmZkZDgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MjlUMDQyOTU2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjMxZjMwMTdlNTQ0NDRiZGQ4NGViZDhkZjMxYWYyNWZkYjJmZGMzMjVjOTk4ZmIxMWFkY2Y2ZWM3ZDNiY2RiNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.P8g5qYGYYl06hq1uKVFgv2r7sp5AkR05OvNTXAFP7bs">
> Thank you for reading and we look forward to your feedback.
> 
> Cheers,
> 
> Team PyTorch

**Comment by fanta1mohamed** [comment]
جيد اريد ان اتعلم

**Comment by atalman** [comment]
Hi @fanta1mohamed please post comments using English to ensure it can be understood and addressed by the development team
